python/ql/src/semmle/python/RegexParserExtended.qll

```diff
@@ -0,0 +1,556 @@
+import Parser
+private import RegexLiteral
+
+module RegexSpecific {
+  predicate allowedEmptyClasses() { none() }
+}
+
+import RegexSpecific as Conf
+
+private string escapableChars() { result = "AbBdDsSwWZafnNrtuUvx\\\\" }
+
+private string keywordChars() { result = "()|*+?\\-\\[\\]" }
```

Comment:
I think you're missing some characters here, e.g. `{` and `}`. I tested the tokenization with the regex `a*b{9}`, and you can see that `tokens` treats the braces as both `normalchar` and as part of `fixedrepeat`:

<img width="439" alt="image" src="https://user-images.githubusercontent.com/158306/119649237-4fab8500-be1a-11eb-91b8-f991cd9e5bb7.png">

But I've only just started to try and understand this code, so it's possible I've misunderstood.

Reply:
`{` and `}` can be a keyword or not based on context.  
E.g. in `a{b` the `{` is a string constant, while `{` is a keyword in `a{2,}`.   
That might be why. 

That is at least the case for JavaScript. 
Even though I think it's in conflict with the specification.  
But everyone seems to do it.

Reply:
Ah, that makes sense.

But I assume we still need to maintain the invariant that each character in a regex literal is only part of one token. So in my example above, the last 3 rows should not be there.

Reply:
> But I assume we still need to maintain the invariant that each character in a regex literal is only part of one token. So in my example above, the last 3 rows should not be there.

We could do that. A regular expression should be able to capture exactly which `{` and `}` are keywords. 
(I haven't dived deep into the implementation, so I might be wrong). 

But I'm not sure we need to do that. 
The way the parser works is that is spuriously creates everything that seems feasible when looking at the very local context. 
And each layer/iteration of the parser then only uses the tokens/ast-nodes that can be used to create a valid ast-node. 

So I don't think spurious tokens are an issue. 

Spurious tokens also arise when parsing dashes inside character classes. 
E.g. in `/foo[A-Z-]bar/`, where the first dash is a keyword, and the second is a constant.  
And I'm not sure if a regular expression can correctly capture which dashes are keywords.